File: Algorithmic_Toolbox_C1/W2_Warm_up/fib_sum_mn_last_num.c
/*
 ============================================================================
 Name        : fib_sum_mn_last_num.c
 Author      : Ivan
 Version     :
 Copyright   : Your copyright notice
 Description : Compute nth fibonacci number mod m Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <time.h>

#define CHAR_BUFFSIZE (int) (1<<20) // 2^20 = 1048576 bytes around 1 mb

int get_fib_sum_last_num_naive(long m, long n) {
	if(n <= 1)
		return n;

	int past = 0;
	int current = 1;
	int temp;
	int sum;
	if(m >= 2){
		sum = 0;
	}else{
		sum = 1;
	}


	for(long i = 2; i <= n; ++i) {
		temp = past;
		past = current;
		current = (current + temp) % 10;

		if(i >= m)
			sum = (sum + current) % 10;
	}

	return sum;
}

char get_fib_sum_last_num_eff(long m, long n){


	char fib_sum[CHAR_BUFFSIZE+1]; fib_sum[0] = 0; fib_sum[1] = 1;
	
	if(n <= 1)
		return (char) n;

	for(long i = 2; i <= n; ++i) {

		fib_sum[i] = (1 + fib_sum[i-1] + fib_sum[i-2]) % 10;

		if(fib_sum[i-1] == 0 && fib_sum[i] == 1){
			if(m == 0){
				return fib_sum[n % (i-1)];
			}else{
			return (10 + ( ( fib_sum[n % (i-1)] - fib_sum[(m-1)%(i-1)] ) % 10 ) ) % 10;
			}
		}
		if(i % CHAR_BUFFSIZE == 0){
			printf("Pisano period is too large or non-existent\n");
			return 0;
		}
	}

	if(m==0){
		return fib_sum[n];
	}else{
		return (10 + ((fib_sum[n] - fib_sum[m-1])%10)) % 10;
	}
}


int stress_test(long max_n){

	long counter = 0;
	long n;
	long m;
	srand(time(0));
    
    while(1){
    	n = rand() % max_n; n++;
    	m = rand() % n; m++;

        int sol1 = get_fib_sum_last_num_naive(m, n);
        int sol2 = get_fib_sum_last_num_eff(m, n);

		char *msg;
		asprintf(&msg, "ERROR: m = %ld n = %ld\n SOL1: %d\nSOL2: %d", m, n, sol1, sol2);
        
        assert(sol1 == sol2 && msg);
        free(msg); // release the memory allocated by asprintf.

        printf("Test %ld successful\n", counter);
        counter = counter + 1;
    }
}

int main(void) {
	int debug = 0;

	if(debug == 1){
		stress_test(5000);
	}
	else{
		long m;
		long n;
		scanf("%ld %ld", &m, &n);
		// printf("%d\n", get_fib_sum_last_num_naive(m, n));
		printf("%d\n", get_fib_sum_last_num_eff(m, n));
	}

	
	return 0;
}

File: Algorithmic_Toolbox_C1/W2_Warm_up/fib_sum_mn_last_num.py
# python3
# Compute the last digit of the sum of Fibonacci numbers in a given range.

import numpy as np
import sys
import time

# Naive implementation, stores whole fibonnaci number, used for stress testing more efficient implementations.
def get_fib_sum_last_digit_naive(m, n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum = 0 if m >= 2 else 1

    for i in range(2, n + 1):
        previous, current = current, (previous + current) % 10
        if i >= m:
            sum = (sum + current) % 10

    return sum

# More efficient implementation, stores only the ones place of each fibonacci number.
def get_fib_sum_last_digit_table(m, n):
    fib_table = [0,1]
    fib_sum = [0, 1]
    if n<=1:
        return n

    for i in range(2,n+1):
        fib_table.append((fib_table[i-1] + fib_table[i-2]))
        if i >= m:
            fib_sum.append(sum(fib_table)%10)

    return fib_sum

# Efficient implementation given that sum of fibonacci numbers mod 10 is a periodic sequence.
def get_fib_sum_last_digit_eff(m, n):
    fib_sum_table = [0, 1]

    if n <= 1:
        return n

    for index in range(2, n+1):
        fib_sum_table.append((1 + fib_sum_table[index - 1] + fib_sum_table[index - 2]) % 10)

        if (fib_sum_table[index - 1] == 0 and fib_sum_table[index] == 1):
            # Return the sum up to n minus the sum up to (m-1). Add 10 and mod by 10 to get correct output.
            if m == 0:
                return fib_sum_table[n % (index - 1)]
            else:
                return (10 + ((fib_sum_table[n % (index - 1)] - fib_sum_table[(m - 1) % (index - 1)]) % 10)) % 10

    if m == 0:
        return fib_sum_table[n]
    else:
        return (10 + ((fib_sum_table[(n)] - fib_sum_table[(m - 1)]) % 10)) % 10


def stress_test(fib_start=20, fib_end=200):
    counter = 0
    while(True):
        m = np.random.randint(fib_start)
        n = np.random.randint(fib_start, fib_end+1)

        sol1 = get_fib_sum_last_digit_naive(m, n)
        sol2 = get_fib_sum_last_digit_eff(m, n)
        assert(sol1 == sol2), f"ERROR: n = {n} m = {m}\n SOL1: {sol1}\nSOL2: {sol2}"
        print(f"Test {counter} succesful")
        counter += 1

if __name__ == '__main__':

    if len(sys.argv) > 1:
        stress_test(int(sys.argv[1]), int(sys.argv[2]))
    else:
        m, n = map(int, input().split())
        # print(get_fib_sq_sum_last_digit_naive(m, n))
        # print(get_fib_sq_sum_last_digit_table(n))
        # start = time.time()
        print(get_fib_sum_last_digit_eff(m, n))
        # print(f"Total time: {time.time()-start}")